VSO CSI driver documentation (#691)
hashicorp/web-unified-docs: content/vault/v1.20.x/content/docs/deploy/kubernetes/comparisons.mdx

Author: @tvoran Would you be able to take a glance at this `comparisons.mdx` page for accuracy, with your knowledge of Injector and CSI Provider? (Feel free to use the vercel preview on the PR to view it in a more visually discernible way) 

A lot of these were populated from older comparison matrices written by other people, so I am not sure how accurate/confusing they are.

  Compare the available Kubernetes integrations for Vault, including Vault Secrets
  Operator, Vault Secrets Store CSI provider,
  and Vault Agent Injector.
---

# Kubernetes integrations comparison

HashiCorp Vault provides three different integrations to help Kubernetes workloads consume secrets from Vault without modifying application code. Each integration has unique characteristics that make it suitable for different use cases.

Use the information below to compare the available integrations and choose the right solution for your needs.

## Features at a glance

| Feature                             | Vault Secrets Operator                                             | Vault Secrets Store CSI provider   | Vault Agent Injector
	| ----------------------------------- | --------------------------------- | ---------------------------- | --------------
	| Supported secret types              | Static/KV, PKI, Dynamic, Database, AppRole secret IDs            | All                  | All
	| Supported auth methods              | K8s, AppRole, GCP, AWS, JWT                                      | K8s, JWT             | K8s, other [auto-auth methods](/vault/docs/agent-and-proxy/autoauth)

> 	| Storage model                       | K8s Secrets or ephemeral volumes with CSI driver                 | Ephemeral volumes    | Ephemeral volumes

Author: Thanks for catching this here; I just learned this myself!

	| Kubernetes-native                   | Yes                                                              | Yes                  | No
	| Secret data persistence             | Persisted in `etcd` or ephemeral with CSI driver                 | Temporarily persisted or ephemeral | Ephemeral
	| Secret data templating              | Yes                                                              | No                   | Yes

> 	| Requires Vault for Pod autoscaling  | No                                                               | Yes                  | Yes

Reviewer: Only in the non CSI case.

	| Secret sharing between applications | Yes                                                              | Yes                  | No

## Vault Secrets Operator

The [Vault Secrets Operator](/vault/docs/platform/k8s/vso) synchronizes secrets from Vault to native Kubernetes Secrets. 

> It operates by watching Custom Resource Definitions (CRDs) and syncing secret data from Vault to Kubernetes Secrets, which applications can consume using standard Kubernetes patterns.

Reviewer: I would replace watching with reconciling

Vault Agent can be deployed alongside the provider to manage [caching and renewals](/vault/docs/agent-and-proxy/agent/caching#caching-and-renewals).

**Best for:** Organizations using multiple secret stores in addition to Vault, or those who prefer vendor-neutral CSI standards.

## Vault Agent Injector

The [Vault Agent Injector](/vault/docs/platform/k8s/injector) injects Vault Agent sidecar containers into pods. 
The Agent containers authenticate with Vault and render secrets to a shared memory volume for consumption by application containers.

**Best for:** Applications that want to reference multiple Vault secrets in one template, or need a wider array of auth methods.

## Use case recommendations

The appropriate integration for your environment depends on the applicable
security policies, secret data governance, ease of use requirements, and
system availability guarantees.

| Limitation                                        | Vault Secrets Operator            | Vault Secrets Store CSI provider　| Vault Agent Injector
| ------------------------------------------------- | --------------------------------- | ---------------------------- | --------------

> | Requires privileged pod access                    | No, except with CSI driver        | Yes                          | No

Reviewer: nit: capitalize Pod as it is a resource like CSISecrets, etc.

| Flexibility in how to consume secrets             | Yes                               | Yes                          | No
| Low resource consumption (I/O, CPU, Memory)       | Yes                               | Yes                          | No
| Secret data limited to ephemeral volumes          | No                                | No                           | Yes

> | Pod Autoscaling dependent on Vault availability   | No                                | Yes                          | Yes

Reviewer: asterix - only in the non CSI case.

| Personas easily separated via K8s [RBAC] policies | Yes                               | No                           | No

[RBAC]: https://kubernetes.io/docs/reference/access-authn-authz/rbac

## Performance considerations

Performance characteristics vary significantly between integrations based on
their architectural patterns.

<Tabs>
	
<Tab heading="Load on Vault">

- **Vault Secrets Operator** - Lowest load due to per-CRD connections and cluster-local caching.
- **Vault Secrets Store CSI provider** - Higher load due to per-pod connections.
- **Vault Agent Injector** - Highest load due to per-pod connections and sidecar pattern.
	
</Tab>
	
<Tab heading="Rendered secret lifecycle">

- **Vault Secrets Operator** - Secret lifecycle is independent of pod lifecycle due 
to caching in Kubernetes Secrets, unless using the Vault Secrets Operator CSI driver.
- **Vault Secrets Store CSI provider** - Secret lifecycle depends on pod lifecycle, even with the 
option to cache in Kubernetes Secrets (they will be deleted after the pod terminates).
- **Vault Agent Injector** - Secret lifecycle depends on pod lifecycle.

</Tab>
	
<Tab heading="Resource consumption">

> - **Vault Secrets Operator** - Lower consumption due to one client process needed per cluster.

Reviewer: In the case of CSI it is one driver instance per Node

Author: Yeah let me add that back in; I'd mentioned that when VSO-CSI was its own bullet point but now we're discussing it as just part of VSO.

- **Vault Secrets Store CSI provider** - Medium consumption due to one client process needed per node.
- **Vault Agent Injector** - Higher consumption due to one client process needed per Pod (as a sidecar container), though resource allocations can be tuned.

</Tab>
	
<Tab heading="Scalability">

Integrations that share secret data across Pods scale well as the number of Pods
increase. Integrations that create independent Vault connections for each Pod do
not scale well as the number of Pods increase.
	
- **Vault Secrets Operator** - shares secret data.
- **Vault Secrets Store CSI provider** - creates independent Vault connections.
- **Vault Agent Injector** - creates independent Vault connections.
	
</Tab>
	
</Tabs>

## Migration requirements 

- Migrating **to** Vault Secrets Operator requires:
    1. Creating CRDs for each relevant secret type. For example, `VaultStaticSecret`, `VaultDynamicSecret`, `VaultPKISecret`, or `CSISecrets`.

>     1. Updating applications to use Kubernetes Secrets.

Reviewer: Would it be worth linking out to the API docs here?

| Pod Autoscaling dependent on Vault availability   | No                                | Yes                          | Yes
| Adoption of the K8s operator model                | Yes                               | Yes                          | No
| Personas easily separated via K8s [RBAC] policies | Yes                               | No                           | No

[RBAC]: https://kubernetes.io/docs/reference/access-authn-authz/rbac

## Performance considerations

Performance characteristics vary significantly between integrations based on
their architectural patterns.

<Tabs>
	
<Tab heading="Load on Vault">

- **Vault Secrets Operator** - Lowest load due to per-CRD connections and cluster-local caching.
- **Vault Secrets Store CSI provider** - Higher load due to per-pod connections.
- **Vault Agent Injector** - Highest load due to per-pod connections and sidecar pattern.
	
</Tab>
	
<Tab heading="Secret availability">

Integrations that use ephemeral storage have variable availability because
secret access depends on Vault availability.

- **Vault Secrets Store CSI provider** - Variable availability.
- **Vault Agent Injector** - Variable availability.

> - **Vault Secrets Operator** - Best availability. Kubernetes Secrets act as durable cache.

Reviewer 2: > secret access depends on Vault availability.

Isn't that true for any of these? i.e. nothing can fetch secrets if Vault is down. If a secret is fetched from Vault and rendered, Vault subsequently going down isn't going to clear the rendered secret for CSI provider or Agent, is it?

Integrations using ephemeral storage means the rendered secret lifecycle is tied to the Pod lifecycle, instead of being independent for VSO, maybe re-frame along those lines?

	
</Tab>
	
<Tab heading="I/O, CPU, and memory consumption">
	
- **Vault Secrets Operator** - Low consumption.
- **Vault Secrets Store CSI provider** - Medium consumption.

> - **Vault Agent Injector** - High consumption due to the number of sidecar containers per pod.

Reviewer 2: It's just one sidecar and/or one init container per Pod, which aren't running simultaneously; maybe just drop the "per pod" bit?

Side note: Do we have data showing high cpu and memory usage for Agent? The default CPU and memory resource requests are pretty high, but they can be tuned down fairly low IIRC.

	| Secret sharing between applications | Yes                                                              | Yes                  | No

## Vault Secrets Operator

The [Vault Secrets Operator](/vault/docs/platform/k8s/vso) synchronizes secrets from Vault to native Kubernetes Secrets. 
It operates by reconciling Custom Resource Definitions (CRDs) and syncing secret data from Vault to Kubernetes Secrets, which applications can consume using standard Kubernetes patterns.
Vault Enterprise users can also configure it to run a [CSI driver](/vault/docs/platform/k8s/vso/csi), which mounts secrets directly from Vault to application containers as volumes, bypassing Kubernetes Secrets. 

**Best for:** Teams who prefer native Kubernetes workflows and want the option of either durable secret storage within the cluster, or written to ephemeral volumes.

## Vault Secrets Store CSI provider

The [Vault Secrets Store CSI provider](/vault/docs/platform/k8s/csi) uses the vendor-neutral [Secrets Store CSI driver](https://secrets-store-csi-driver.sigs.k8s.io/) to mount secrets from Vault as ephemeral volumes. 

> Vault Agent can be deployed alongside the provider to manage [caching and renewals](/vault/docs/agent-and-proxy/agent/caching#caching-and-renewals).

Reviewer 3: ```suggestion
You can deploy Vault Agent alongside the provider to manage [caching and renewals](/vault/docs/agent-and-proxy/agent/caching#caching-and-renewals).
```
Style correction: write in active voice

**Best for:** Organizations using multiple secret stores in addition to Vault, or those who prefer vendor-neutral CSI standards.

## Vault Agent Injector

The [Vault Agent Injector](/vault/docs/platform/k8s/injector) injects Vault Agent sidecar containers into Pods. 
The Agent containers authenticate with Vault and render secrets to a shared memory volume for consumption by application containers.

> **Best for:** Applications that want to reference multiple Vault secrets in one template, or need a wider array of auth methods.

Reviewer 3: ```suggestion
**Best for:** Applications that want to reference multiple Vault secrets in one template or need a wider array of auth methods.
```

## Use case recommendations

The appropriate integration for your environment depends on the applicable
security policies, secret data governance, ease of use requirements, and
system availability guarantees.

> | Limitation                                        | Vault Secrets Operator            | Vault Secrets Store CSI provider　| Vault Agent Injector
> | ------------------------------------------------- | --------------------------------- | ---------------------------- | --------------
> | Requires privileged Pod access                    | No, except with CSI driver        | Yes                          | No
> | Flexibility in how to consume secrets             | Yes                               | Yes                          | No
> | Low resource consumption (I/O, CPU, Memory)       | Yes                               | Yes                          | No
> | Secret data limited to ephemeral volumes          | No                                | No                           | Yes
> | Pod Autoscaling dependent on Vault availability   | No, except with CSI driver        | Yes                          | Yes
> | Personas easily separated via K8s [RBAC] policies | Yes                               | No                           | No

Reviewer 3: ```suggestion
| Limitation                                        | Vault Secrets Operator              | Vault Secrets Store CSI provider　| Vault Agent Injector
| ------------------------------------------------- | ----------------------------------- | ---------------------------- | --------------
| Requires privileged Pod access                    | No (default) <br/> Yes (CSI driver) | Yes                          | No
| Flexibility in how to consume secrets             | Yes                                 | Yes                          | No
| Low resource consumption (I/O, CPU, Memory)       | Yes                                 | Yes                          | No
| Secret data limited to ephemeral volumes          | No                                  | No                           | Yes
| Pod Autoscaling dependent on Vault availability   | No (default) <br/> Yes (CSI driver) | Yes                          | Yes
| Personas easily separated via K8s [RBAC] policies | Yes                                 | No                           | No
```

[RBAC]: https://kubernetes.io/docs/reference/access-authn-authz/rbac

## Performance considerations

Performance characteristics vary significantly between integrations based on
their architectural patterns.

<Tabs>
	
<Tab heading="Load on Vault">

- **Vault Secrets Operator** - Lowest load due to per-CRD connections and cluster-local caching. 
Single manager for an entire cluster. If using the CSI driver, there will be one Vault connection pool per Node.
- **Vault Secrets Store CSI provider** - Higher load due to per-Pod connections.
- **Vault Agent Injector** - Highest load due to per-Pod connections and sidecar pattern.
	
</Tab>
	
<Tab heading="Rendered secret lifecycle">

> - **Vault Secrets Operator** - Secret lifecycle is independent of Pod lifecycle due 

Reviewer 3: Break this into two lines

to caching in Kubernetes Secrets, unless using the Vault Secrets Operator CSI driver.

> - **Vault Secrets Store CSI provider** - Secret lifecycle depends on Pod lifecycle, even with the 
> option to cache in Kubernetes Secrets (they will be deleted after the Pod terminates).

Reviewer 3: ```suggestion
- **Vault Secrets Store CSI provider** - Secret lifecycle depends on Pod lifecycle, even with the 
option to cache in Kubernetes Secrets (they will be deleted after the Pod terminates).```suggestion
- **Vault Secrets Store CSI provider** - Secret lifecycle depends on Pod
  lifecycle as Vault deletes cached Kubernetes Secrets when the Pod terminates.
```
Style correction: write in active voice

- **Vault Agent Injector** - Secret lifecycle depends on Pod lifecycle.

</Tab>
	
<Tab heading="Resource consumption">

> - **Vault Secrets Operator** - Lowest consumption due to one client process needed per Kubernetes cluster. 

Reviewer 3: Break this into two lines

increase. Integrations that create independent Vault connections for each Pod do
not scale well as the number of Pods increase.
	
- **Vault Secrets Operator** - shares secret data.
- **Vault Secrets Store CSI provider** - creates independent Vault connections.
- **Vault Agent Injector** - creates independent Vault connections.
	
</Tab>
	
</Tabs>

## Migrating to Vault Secrets Operator 

Migrating to Vault Secrets Operator from the other Kubernetes integrations requires:

1. Creating CRDs for each relevant secret type, such as `VaultStaticSecret`, `VaultDynamicSecret`, `VaultPKISecret`, or `CSISecrets`. Refer to the [API reference](/vault/docs/platform/k8s/vso/api-reference) for configuration details.
Note that the Vault Secrets Operator CSI driver only supports KV secrets and AppRole secret IDs.
1. Updating applications to use Kubernetes Secrets.
1. Migrating from Vault Agent Injector may require updating your secret templates to 
use Vault Secrets Operator's `SecretTransformation` resource. 
While both tools use Go templating, Vault Agent Injector templates may not work with the `SecretTransformation` resource out of the box.

> 1. If using the [direct secrets](/vault/docs/platform/k8s/vso/csi) functionality, you will also need:
>   1. A Vault Enterprise server.
>   1. Updated Pod specifications that reference CSI volumes.

Reviewer 3: ```suggestion
1. To use [direct secrets](/vault/docs/platform/k8s/vso/csi) you also need:
  1. A Vault Enterprise license.
  1. Updated Pod specifications that reference CSI volumes.
```
Style correction: write in active voice

| Adoption of the K8s operator model                | Yes                               | Yes                          | No
| Personas easily separated via K8s [RBAC] policies | Yes                               | No                           | No

[RBAC]: https://kubernetes.io/docs/reference/access-authn-authz/rbac

## Performance considerations

Performance characteristics vary significantly between integrations based on
their architectural patterns.

<Tabs>
	
<Tab heading="Load on Vault">
	
- **Vault Secrets Operator** - Lowest load due to per-CRD connections and
  cluster-local caching.
- **Vault Secrets Store CSI provider** - Higher load due to per-pod connections.
- **Vault Agent Injector** - Highest load due to per-pod connections and sidecar
  pattern.
	
</Tab>
	
<Tab heading="Secret availability">
	
Integrations that use ephemeral storage have variable availability because
secret access depends on Vault availability.
	
- **Vault Secrets Store CSI provider** - Variable availability.
- **Vault Agent Injector** - Variable availability.
- **Vault Secrets Operator** - Best availability. Kubernetes Secrets act
  as durable cache.
	
</Tab>
	

> <Tab heading="I/O, CPU, and memory consumption">

Reviewer 2: Maybe just "Resource consumption"?

	
- **Vault Secrets Operator** - Low consumption.

> - **Vault Secrets Store CSI provider** - Medium consumption.

Reviewer 2: What are the reasons for the difference in consumption for these two?

> <Tabs>
> 	
> 	<Tab heading="Load on Vault">
> 	
> 	- **Vault Secrets Operator** - Lowest load due to per-CRD connections and
> 	   cluster-local caching.
> 	- **Vault Secrets Store CSI provider** - Higher load due to per-pod connections.
> 	- **Vault Agent Injector** - Highest load due to per-pod connections and sidecar
> 	   pattern.
> 	
> 	</Tab>
> 	
> 	<Tab heading="Secret availability">
> 	
> 	Integrations that use ephemeral storage have variable availability because
> 	secret access depends on Vault availability.
> 	
> 	- **Vault Secrets Store CSI provider** - Variable availability.
> 	- **Vault Agent Injector** - Variable availability.
> 	- **Vault Secrets Operator** - Best availability. Kubernetes Secrets act
> 	  as durable cache.
> 	
> 	</Tab>
> 	
> 	<Tab heading="I/O, CPU, and memory consumption">
> 	
> 	- **Vault Secrets Operator** - Low consumption.
> 	- **Vault Secrets Store CSI provider** - Medium consumption.
> 	- **Vault Agent Injector** - High consumption due to the number of sidecar containers
> 	  per pod.
> 	
> 	</Tab>
> 	
> 	<Tab heading="Scalability">
> 	
> 	Integrations that share secret data across pods scale well as the number of pods
> 	increase. Integrations that create independent Vault connections for each pod do
> 	not scale well as the number of pods increase.
> 	
> 	- **Vault Secrets Operator** - shares secret data.
> 	- **Vault Secrets Store CSI provider** - creates independent Vault connections.
> 	- **Vault Agent Injector** - creates independent Vault connections.
> 	
> 	</Tab>
> 	
> 	</Tabs>

Reviewer 3: ```suggestion
<Tabs>
	
<Tab heading="Load on Vault">
	
- **Vault Secrets Operator** - Lowest load due to per-CRD connections and
  cluster-local caching.
- **Vault Secrets Store CSI provider** - Higher load due to per-pod connections.
- **Vault Agent Injector** - Highest load due to per-pod connections and sidecar
  pattern.
	
</Tab>
	
<Tab heading="Secret availability">
	
Integrations that use ephemeral storage have variable availability because
secret access depends on Vault availability.
	
- **Vault Secrets Store CSI provider** - Variable availability.
- **Vault Agent Injector** - Variable availability.
- **Vault Secrets Operator** - Best availability. Kubernetes Secrets act
  as durable cache.
	
</Tab>
	
<Tab heading="I/O, CPU, and memory consumption">
	
- **Vault Secrets Operator** - Low consumption.
- **Vault Secrets Store CSI provider** - Medium consumption.
- **Vault Agent Injector** - High consumption due to the number of sidecar containers
  per pod.
	
</Tab>
	
<Tab heading="Scalability">

Integrations that share secret data across pods scale well as the number of pods
increase. Integrations that create independent Vault connections for each pod do
not scale well as the number of pods increase.
	
- **Vault Secrets Operator** - shares secret data.
- **Vault Secrets Store CSI provider** - creates independent Vault connections.
- **Vault Agent Injector** - creates independent Vault connections.
	
</Tab>
	
</Tabs>
```
